coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -461,49 +475,61 @@ class SingleClusterPlanner(val dataset: Dataset,
       case aif: ApplyInstantFunction => helper(aif.vectors)
       case bj: BinaryJoin =>
         // lhs/rhs must reside on the same set of shards, and target schema labels for all leaves must be
-        //   discoverable, equal, and preserved by join keys
-        val lhsShards = helper(bj.lhs)
-        val rhsShards = helper(bj.rhs)
-        val canPushdown = lhsShards != None && rhsShards != None &&
-                          // either the shard groups are equal, or either of lhs/rhs includes only scalars.
-                          (lhsShards == rhsShards || lhsShards.get.isEmpty || rhsShards.get.isEmpty) &&
-                          {
-                            val targetSchemaLabels =
-                              getUniversalTargetSchemaLabels(bj, targetSchemaProvider(qContext))
-                            targetSchemaLabels.isDefined &&
-                              targetSchemaLabels.get.toSet.diff(shardColumns.toSet).subsetOf(bj.on.toSet)
-                          }
-        // union lhs/rhs shards, since one might be empty (if it's a scalar)
-        if (canPushdown) Some(lhsShards.get.union(rhsShards.get)) else None
+        //   discoverable, equal, and preserved by join keys.
+        val lhsData = helper(bj.lhs)
+        val rhsData = helper(bj.rhs)
+        val canPushdown = lhsData.isDefined && rhsData.isDefined && {
+          val hasScalar = lhsData.get.shards.isEmpty || rhsData.get.shards.isEmpty
+          val canPushdownJoin = lhsData.get.shards == rhsData.get.shards &&
```

Comment:
Is the order of ``Seq`` containing shards granted to be ordered consistently? ``Seq(1, 2, 3) != Seq(3, 2, 1)`` though you want to push down.

Also too much going on here, some comments would really be helpful to understand the intent.